extensions/standard-processors/processors/ReplaceText.cpp

```diff
@@ -0,0 +1,339 @@
+  gsl_Expects(flow_file);
+  gsl_Expects(session);
+
+  try {
+    utils::LineByLineInputOutputStreamCallback read_write_callback{[this, &flow_file, &parameters](const std::string& input_line, bool is_first_line, bool is_last_line) {
+      switch (line_by_line_evaluation_mode_.value()) {
+        case LineByLineEvaluationModeType::ALL:
+          return applyReplacements(input_line, flow_file, parameters);
+        case LineByLineEvaluationModeType::FIRST_LINE:
+          return is_first_line ? applyReplacements(input_line, flow_file, parameters) : input_line;
+        case LineByLineEvaluationModeType::LAST_LINE:
+          return is_last_line ? applyReplacements(input_line, flow_file, parameters) : input_line;
+        case LineByLineEvaluationModeType::EXCEPT_FIRST_LINE:
+          return is_first_line ? input_line : applyReplacements(input_line, flow_file, parameters);
+        case LineByLineEvaluationModeType::EXCEPT_LAST_LINE:
+          return is_last_line ? input_line: applyReplacements(input_line, flow_file, parameters);
+      }
+      throw Exception{PROCESSOR_EXCEPTION, utils::StringUtils::join_pack("Unsupported ", LineByLineEvaluationMode.getName(), ": ", line_by_line_evaluation_mode_.toString())};
+    }};
+    session->readWrite(flow_file, &read_write_callback);
+    session->transfer(flow_file, Success);
+  } catch (const Exception& exception) {
+    logger_->log_error("Error in ReplaceText (Line-by-Line mode): %s", exception.what());
+    session->transfer(flow_file, Failure);
+  }
+}
+
+std::string ReplaceText::applyReplacements(const std::string& input, const std::shared_ptr<core::FlowFile>& flow_file, const Parameters& parameters) const {
+  const auto [chomped_input, line_ending] = utils::StringUtils::chomp(input);
+
+  switch (replacement_strategy_.value()) {
+    case ReplacementStrategyType::PREPEND:
+      return parameters.replacement_value_ + input;
+
+    case ReplacementStrategyType::APPEND:
+      return chomped_input + parameters.replacement_value_ + line_ending;
+
+    case ReplacementStrategyType::REGEX_REPLACE:
+      return std::regex_replace(chomped_input, parameters.search_regex_, parameters.replacement_value_) + line_ending;
+
+    case ReplacementStrategyType::LITERAL_REPLACE:
+      return applyLiteralReplace(chomped_input, parameters) + line_ending;
+
+    case ReplacementStrategyType::ALWAYS_REPLACE:
+      return parameters.replacement_value_ + line_ending;
+
+    case ReplacementStrategyType::SUBSTITUTE_VARIABLES:
+      return applySubstituteVariables(chomped_input, flow_file) + line_ending;
+  }
+
+  throw Exception{PROCESSOR_EXCEPTION, utils::StringUtils::join_pack("Unsupported ", ReplacementStrategy.getName(), ": ", replacement_strategy_.toString())};
+}
+
+std::string ReplaceText::applyLiteralReplace(const std::string& input, const Parameters& parameters) const {
```

Comment:
`applyLiteralReplace` could be `static`, or a free utility function in an anonymous namespace.

Reply:
Yes it could, but should it?  We would be marking this function as different from those around it, when the difference is only "this function doesn't currently log anything", which is a minor implementation detail.  The only advantage I can see of adding "static" would be to get rid of the clang-tidy warning, but I don't think we should let clang-tidy push us around like that. :)

Reply:
I see this as an unnecessary dependency just for subjective aesthetics. I also don't want to block the PR because of this minor detail if you disagree.

I like following clang-tidy suggestions unless I have a good argument against it.

Reply:
I think pretty patterns are useful, too, as they improve readability.  Also, I suspect it won't take long before cpplint and clang-tidy disagree about something and then we'll be stuck.  But okay: https://github.com/apache/nifi-minifi-cpp/pull/1170/commits/4b1476c18104f4fa9de3994c9b366c284177e779.

Reply:
I don't want to force my opinion on you, so feel free to revert the change and close this thread if I didn't convince you. I think minor issues like this are well within the author's discretion.